src/cosmos-copilot.WebApp/Services/ChatService.cs

```diff
@@ -151,41 +145,21 @@ public async Task<Message> GetChatCompletionAsync(string tenantId, string userId
     /// </summary>
     private async Task<List<Message>> GetChatSessionContextWindow(string tenantId, string userId, string sessionId)
     {
-        ArgumentNullException.ThrowIfNull(tenantId);
-        ArgumentNullException.ThrowIfNull(userId);
-        ArgumentNullException.ThrowIfNull(sessionId);
-
-        int? tokensUsed = 0;
-
+        
         List<Message> allMessages = await _cosmosDbService.GetSessionMessagesAsync(tenantId, userId, sessionId);
-        List<Message> contextWindow = new List<Message>();
-
-        //Start at the end of the list and work backwards
+ 
+        //Build the contextWindow from allMessages, start from the end and work backwards
         //This includes the latest user prompt which is already cached
-        for (int i = allMessages.Count - 1; i >= 0; i--)
-        {
-            tokensUsed += allMessages[i].PromptTokens + allMessages[i].CompletionTokens;
-
-            if (tokensUsed > _maxConversationTokens)
-                break;
+        
+        return allMessages.TakeLast(_maxContextWindow).ToList();
```

Comment:
Instead of taking the last few messages up to `_maxContextWindow` should we just change the query in `GetSessionMessagesAsync()` from the Cosmos service to pull the X most recent chats from the conversation?

Reply:
Yes, that is definitely more efficient for a call that is made frequently. However, changing CosmosService.GetSessionMessagesAsync() impacts ChatService.SummarizeChatSessionNameAsync() so you'd need to rewrite so that both calls work.  Go ahead if you want to do that.